cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/PreventBrokerScaleDownUtils.java

```diff
@@ -0,0 +1,239 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.operator.cluster.operator.assembly;
+
+import io.strimzi.api.kafka.model.KafkaResources;
+import io.strimzi.api.kafka.model.StrimziPodSet;
+import io.strimzi.api.kafka.model.status.KafkaStatus;
+import io.strimzi.operator.cluster.model.KafkaCluster;
+import io.strimzi.operator.cluster.model.NodeRef;
+import io.strimzi.operator.cluster.model.PodSetUtils;
+import io.strimzi.operator.common.AdminClientProvider;
+import io.strimzi.operator.common.Reconciliation;
+import io.strimzi.operator.common.ReconciliationLogger;
+import io.strimzi.operator.common.operator.resource.SecretOperator;
+import io.strimzi.operator.common.operator.resource.StatusUtils;
+import io.strimzi.operator.common.operator.resource.StrimziPodSetOperator;
+import io.vertx.core.Future;
+import io.vertx.core.Promise;
+import io.vertx.core.Vertx;
+import org.apache.kafka.clients.admin.Admin;
+import org.apache.kafka.clients.admin.ListTopicsOptions;
+import org.apache.kafka.clients.admin.TopicDescription;
+import org.apache.kafka.common.KafkaException;
+import org.apache.kafka.common.TopicPartitionInfo;
+
+import java.util.Collection;
+import java.util.Set;
+import java.util.List;
+import java.util.Map;
+import java.util.HashSet;
+import java.util.ArrayList;
+import static java.util.stream.Collectors.toList;
+
+/**
+ * Class which contains several utility function which check if broker scale down can be done or not.
+ */
+public class PreventBrokerScaleDownUtils {
+
+    /* logger*/
+    private static final ReconciliationLogger LOGGER = ReconciliationLogger.create(KafkaReconciler.class.getName());
+
+    /**
+     * Checks if broker contains any partition replicas when scaling down
+     *
+     * @param vertx                Instance of vertx
+     * @param reconciliation       Reconciliation marker
+     * @param kafka                Instance of Kafka Cluster
+     * @param kafkaStatus          The Kafka Status class for adding conditions to it during the reconciliation
+     * @param secretOperator       Secret operator for working with Secrets
+     * @param adminClientProvider  Used to create the Admin client instance
+     * @param strimziPodSetOperator podSet operator
+     * @return  Future which completes when the check is complete
+     */
+    public static Future<Void> canScaleDownBrokers(Vertx vertx, Reconciliation reconciliation, KafkaCluster kafka, KafkaStatus kafkaStatus,
+                                                   SecretOperator secretOperator, AdminClientProvider adminClientProvider, StrimziPodSetOperator strimziPodSetOperator) {
+
+
+        Set<String> desiredPodNames = new HashSet<>();
+        for (NodeRef node : kafka.nodes()) {
+            desiredPodNames.add(node.podName());
+        }
+
+        return strimziPodSetOperator.listAsync(reconciliation.namespace(), kafka.getSelectorLabels())
```

Comment:
Can we avoid listing the pod sets once again here? The ID assignment classes should already know which brokers are being scaled down, or?